### ISSUE TITLE
Failing test: `test__author_queried_object_fix`

### ISSUE THREAD

**Issue by philipjohn** [Automattic/Co-Authors-Plus#398]
> There was 1 failure:
> 
> 1) Test_Author_Queried_Object::test__author_queried_object_fix
> is_archive, is_author is expected to be true. is_404 is expected to be false.
> Failed asserting that false is true.
> 
> /srv/www/wordpress-develop/tests/phpunit/includes/testcase.php:616
> /srv/www/wpmu-subdirectory/wp-content/plugins/co-authors-plus/tests/test-author-queried-object.php:73

This test has been failing since it's introduction. It fails at this point:

```
// Remove the user from the blog
remove_user_from_blog( $author2, $blog2 );

/**
 * Author 2 was removed from the blog, but still a published author
 */
$this->go_to( get_author_posts_url( $author2 ) );
$this->assertQueryTrue( 'is_author', 'is_archive' );
```

The issue is that `is_404` is true. It should be false, because the author archive page should still display one post.

A manual check of the flow in this test is successful - the author archive displays the one post as expected.

As an additional check I deleted the user from the blog using `remove_user_from_blog()` via WP CLI instead of through the admin, just to check if there was a difference. There was none.

**Comment by philipjohn** [comment]
There's a bunch of DB errors:

> WordPress database error Can't reopen table: 'wptests_2_term_relationships' for query SELECT SQL_CALC_FOUND_ROWS  wptests_2_posts.ID FROM wptests_2_posts  LEFT JOIN wptests_2_term_relationships ON (wptests_2_posts.ID = wptests_2_term_relationships.object_id) LEFT JOIN wptests_2_term_relationships AS tr1 ON (wptests_2_posts.ID = tr1.object_id) LEFT JOIN wptests_2_term_taxonomy ON ( tr1.term_taxonomy_id = wptests_2_term_taxonomy.term_taxonomy_id ) WHERE 1=1  AND ((wptests_2_posts.post_author = 2 OR (wptests_2_term_taxonomy.taxonomy = 'author' AND wptests_2_term_taxonomy.term_id = '2'))) AND wptests_2_posts.post_type = 'post' AND (wptests_2_posts.post_status = 'publish') GROUP BY wptests_2_posts.ID HAVING MAX( IF ( wptests_2_term_taxonomy.taxonomy = 'author', IF (  wptests_2_term_taxonomy.term_id = '2',2,1 ),0 ) ) <> 1  ORDER BY wptests_2_posts.post_date DESC LIMIT 0, 10 made by PHPUnit_TextUI_Command::main, PHPUnit_TextUI_Command->run, PHPUnit_TextUI_TestRunner->doRun, PHPUnit_Framework_TestSuite->run, PHPUnit_Framework_TestSuite->run, PHPUnit_Framework_TestCase->run, PHPUnit_Framework_TestResult->run, PHPUnit_Framework_TestCase->runBare, PHPUnit_Framework_TestCase->runTest, ReflectionMethod->invokeArgs, Test_Author_Queried_Object->test__author_queried_object_fix, WP_UnitTestCase->go_to, WP->main, WP->query_posts, WP_Query->query, WP_Query->get_posts
> WordPress database error Can't reopen table: 'wptests_2_term_relationships' for query SELECT SQL_CALC_FOUND_ROWS  wptests_2_posts.ID FROM wptests_2_posts  LEFT JOIN wptests_2_term_relationships ON (wptests_2_posts.ID = wptests_2_term_relationships.object_id) LEFT JOIN wptests_2_term_relationships AS tr1 ON (wptests_2_posts.ID = tr1.object_id) LEFT JOIN wptests_2_term_taxonomy ON ( tr1.term_taxonomy_id = wptests_2_term_taxonomy.term_taxonomy_id ) WHERE 1=1  AND (wptests_2_posts.post_author = 3) AND wptests_2_posts.post_type = 'post' AND (wptests_2_posts.post_status = 'publish') GROUP BY wptests_2_posts.ID HAVING MAX( IF ( wptests_2_term_taxonomy.taxonomy = 'author', IF (  wptests_2_term_taxonomy.term_id = '2',2,1 ),0 ) ) <> 1  ORDER BY wptests_2_posts.post_date DESC LIMIT 0, 10 made by PHPUnit_TextUI_Command::main, PHPUnit_TextUI_Command->run, PHPUnit_TextUI_TestRunner->doRun, PHPUnit_Framework_TestSuite->run, PHPUnit_Framework_TestSuite->run, PHPUnit_Framework_TestCase->run, PHPUnit_Framework_TestResult->run, PHPUnit_Framework_TestCase->runBare, PHPUnit_Framework_TestCase->runTest, ReflectionMethod->invokeArgs, Test_Author_Queried_Object->test__author_queried_object_fix, WP_UnitTestCase->go_to, WP->main, WP->query_posts, WP_Query->query, WP_Query->get_posts
> WordPress database error Can't reopen table: 'wptests_2_term_relationships' for query SELECT SQL_CALC_FOUND_ROWS  wptests_2_posts.ID FROM wptests_2_posts  LEFT JOIN wptests_2_term_relationships ON (wptests_2_posts.ID = wptests_2_term_relationships.object_id) LEFT JOIN wptests_2_term_relationships AS tr1 ON (wptests_2_posts.ID = tr1.object_id) LEFT JOIN wptests_2_term_taxonomy ON ( tr1.term_taxonomy_id = wptests_2_term_taxonomy.term_taxonomy_id ) WHERE 1=1  AND (wptests_2_posts.post_author = 3) AND wptests_2_posts.post_type = 'post' AND (wptests_2_posts.post_status = 'publish') GROUP BY wptests_2_posts.ID HAVING MAX( IF ( wptests_2_term_taxonomy.taxonomy = 'author', IF (  wptests_2_term_taxonomy.term_id = '2',2,1 ),0 ) ) <> 1  ORDER BY wptests_2_posts.post_date DESC LIMIT 0, 10 made by PHPUnit_TextUI_Command::main, PHPUnit_TextUI_Command->run, PHPUnit_TextUI_TestRunner->doRun, PHPUnit_Framework_TestSuite->run, PHPUnit_Framework_TestSuite->run, PHPUnit_Framework_TestCase->run, PHPUnit_Framework_TestResult->run, PHPUnit_Framework_TestCase->runBare, PHPUnit_Framework_TestCase->runTest, ReflectionMethod->invokeArgs, Test_Author_Queried_Object->test__author_queried_object_fix, WP_UnitTestCase->go_to, WP->main, WP->query_posts, WP_Query->query, WP_Query->get_posts
> WordPress database error Can't reopen table: 'wptests_2_term_relationships' for query SELECT SQL_CALC_FOUND_ROWS  wptests_2_posts.ID FROM wptests_2_posts  LEFT JOIN wptests_2_term_relationships ON (wptests_2_posts.ID = wptests_2_term_relationships.object_id) LEFT JOIN wptests_2_term_relationships AS tr1 ON (wptests_2_posts.ID = tr1.object_id) LEFT JOIN wptests_2_term_taxonomy ON ( tr1.term_taxonomy_id = wptests_2_term_taxonomy.term_taxonomy_id ) WHERE 1=1  AND ((wptests_2_posts.post_author = 3 OR (wptests_2_term_taxonomy.taxonomy = 'author' AND wptests_2_term_taxonomy.term_id = '3'))) AND wptests_2_posts.post_type = 'post' AND (wptests_2_posts.post_status = 'publish') GROUP BY wptests_2_posts.ID HAVING MAX( IF ( wptests_2_term_taxonomy.taxonomy = 'author', IF (  wptests_2_term_taxonomy.term_id = '3',2,1 ),0 ) ) <> 1  ORDER BY wptests_2_posts.post_date DESC LIMIT 0, 10 made by PHPUnit_TextUI_Command::main, PHPUnit_TextUI_Command->run, PHPUnit_TextUI_TestRunner->doRun, PHPUnit_Framework_TestSuite->run, PHPUnit_Framework_TestSuite->run, PHPUnit_Framework_TestCase->run, PHPUnit_Framework_TestResult->run, PHPUnit_Framework_TestCase->runBare, PHPUnit_Framework_TestCase->runTest, ReflectionMethod->invokeArgs, Test_Author_Queried_Object->test__author_queried_object_fix, WP_UnitTestCase->go_to, WP->main, WP->query_posts, WP_Query->query, WP_Query->get_posts
> WordPress database error Can't reopen table: 'wptests_2_term_relationships' for query SELECT SQL_CALC_FOUND_ROWS  wptests_2_posts.ID FROM wptests_2_posts  LEFT JOIN wptests_2_term_relationships ON (wptests_2_posts.ID = wptests_2_term_relationships.object_id) LEFT JOIN wptests_2_term_relationships AS tr1 ON (wptests_2_posts.ID = tr1.object_id) LEFT JOIN wptests_2_term_taxonomy ON ( tr1.term_taxonomy_id = wptests_2_term_taxonomy.term_taxonomy_id ) WHERE 1=1  AND ((wptests_2_posts.post_author = 3 OR (wptests_2_term_taxonomy.taxonomy = 'author' AND wptests_2_term_taxonomy.term_id = '3'))) AND wptests_2_posts.post_type = 'post' AND (wptests_2_posts.post_status = 'publish') GROUP BY wptests_2_posts.ID HAVING MAX( IF ( wptests_2_term_taxonomy.taxonomy = 'author', IF (  wptests_2_term_taxonomy.term_id = '3',2,1 ),0 ) ) <> 1  ORDER BY wptests_2_posts.post_date DESC LIMIT 0, 10 made by PHPUnit_TextUI_Command::main, PHPUnit_TextUI_Command->run, PHPUnit_TextUI_TestRunner->doRun, PHPUnit_Framework_TestSuite->run, PHPUnit_Framework_TestSuite->run, PHPUnit_Framework_TestCase->run, PHPUnit_Framework_TestResult->run, PHPUnit_Framework_TestCase->runBare, PHPUnit_Framework_TestCase->runTest, ReflectionMethod->invokeArgs, Test_Author_Queried_Object->test__author_queried_object_fix, WP_UnitTestCase->go_to, WP->main, WP->query_posts, WP_Query->query, WP_Query->get_posts
>

**Comment by philipjohn** [comment]
May be something to do with [referring to the same temporary table more than once in a query](http://stackoverflow.com/questions/343402/getting-around-mysql-cant-reopen-table-error).

**Comment by trepmal** [comment]
The temporary tables are due to the way WordPress gets installed on the fly for testing.

https://core.trac.wordpress.org/browser/trunk/tests/phpunit/includes/testcase.php#L312 - summarized:

```
add_filter( 'query', function( $query ) {
	if ( 'CREATE TABLE' === substr( trim( $query ), 0, 12 ) ) {
		return substr_replace( trim( $query ), 'CREATE TEMPORARY TABLE', 0, 12 );
	}
	return $query;
} );

```

Interestingly (or not), we're only encountering this because we test in multisite. The initial tables that are installed are not temporary, but the subsite tables are (since those pass through the `query` filter).

If I simply comment out those filters and rerun the tests, they pass.

**Comment by philipjohn** [comment]
Fixed y #457

**Comment by TheCrowned** [comment]
Hi there,
So it looks like the fix to the tables being temporary is giving me issues running tests @trepmal.

When I run the tests the first time, they all succeed. The problem is, if I run them another time, then the following error is displayed:

> `<h1>Already Installed</h1><p>You appear to have already installed WordPress. To reinstall please clear your old database tables first.</p>`

![schermata del 2018-05-30 21-48-44](https://user-images.githubusercontent.com/7880569/40768690-5073f4ec-64b6-11e8-98dc-553cd8c3e353.png)

I traced it down being generated by `test__author_queried_object_fix`, and only when running tests multisite. I have read that after tests are run, the database should be properly cleaned by the WP testcase itself, so I dug into the core tests handling to see why that was not happening. All queries were being run as `TEMPORARY`, except the ones from this test. 

This is a problem, because this statement (which is where the test hangs for me):

```php
$blog2 = $this->factory->blog->create( array( 'user_id' => $author1 ) );
```

will fail if the blog is already present.

And indeed, it fails because of the filters in `test-author-queried-object.php` `setUp()` function, which prevents the default behavior of declaring the new site tables as `TEMPORARY` from happening. The thing is, **if I comment those two filters out, tests succeed AND can be run an arbitrary number of times (without throwing any errors)**.

```php
function setUp() {
	parent::setUp();

	//remove_filter( 'query', array( $this, '_create_temporary_tables' ) );
	//remove_filter( 'query', array( $this, '_drop_temporary_tables' ) );
}
```

Are you sure those filters should be there? `WP_UnitTestCase->tearDown()`, which executes `$wpdb->query( 'ROLLBACK' );` and thus deletes all temporary tables, should be called at the _end of each plugin test_, so it shouldn't create any issues...

Maybe something has changed in these months (although I checked [here](https://core.trac.wordpress.org/browser/trunk/tests/phpunit/includes/factory/class-wp-unittest-factory-for-blog.php?annotate=blame) and [here](https://core.trac.wordpress.org/log/trunk/tests/phpunit/includes/testcase.php) and could not find anything)? Maybe I am doing something wrong? :)

(Quick final note: just to confirm that it is indeed an issue with the multisite-ness of the tests, if I run them not multisite they all succeed (except `test__author_queried_object_fix`, of course), and work multiple times.)
